governator-legacy/src/test/java/com/netflix/governator/guice/TestLifecycleInjector.java

```diff
@@ -11,98 +12,324 @@
+  public void testSingletonProviderPreDestroy() throws Exception {
+    checkCleanup(new AbstractModule() {
+      @Override
+      protected void configure() {
+
+      }
+
+      @Provides
+      @Singleton
+      public PreDestroyOnce providesObj(String name) {
+        return new PreDestroyOnceImpl();
+      }
+    }, JavaxSingletonConsumer.class);
+  }
+
+  /**
+   * registers ProviderMethodProviderInstanceBinding in 'No Scope' 1000 times
+   */
+  @Test
+  public void testUnscopedProviderPreDestroy() throws Exception {
+    checkCleanup(new AbstractModule() {
+      @Override
+      protected void configure() {
+
       }
-    });
+
+      @Provides
+      public PreDestroyOnce providesObj(String name) {
+        return new PreDestroyOnceImpl();
+      }
+    }, JavaxSingletonConsumer.class);
   }
 
+ /**
+   * registers LinkedProviderBinding in 'No Scope' 1000 times
+   */
+   @Test
+  public void testOptionalPreDestroy() throws Exception {
+    checkCleanup(new AbstractModule() {
+      @Override
+      protected void configure() {
+        OptionalBinder.newOptionalBinder(binder(), PreDestroyOnce.class).setDefault()
+            .toProvider(PreDestroyOnceProvider.class);
+      }
+
+    }, JavaxOptionalConsumer.class);
+  }
+
+  /**
+   * registers LinkedProviderBinding in Singleton scope 1 time
+   */
+  @Test
+  public void testSingletonOptionalPreDestroy() throws Exception {
+    checkCleanup(new AbstractModule() {
+      @Override
+      protected void configure() {
+        OptionalBinder.newOptionalBinder(binder(), PreDestroyOnce.class).setDefault()
+            .toProvider(PreDestroyOnceProvider.class).asEagerSingleton();
+      }
+
+    }, JavaxOptionalConsumer.class);
+  }
+
+
   private interface PreDestroyOnce {
     boolean isClosed();
+
+    int getCount();
+  }
+
+  private static class JavaxSingletonConsumer implements PreDestroyOnce {
+    private final PreDestroyOnce delegate;
+
+    @Inject
+    JavaxSingletonConsumer(Provider<PreDestroyOnce> provider) {
+      this.delegate = provider.get();
+    }
+
+    public boolean isClosed() {
+      return delegate.isClosed();
+    }
+
+    public int getCount() {
+      return delegate.getCount();
+    }
+  }
+
+  private static class JavaxOptionalConsumer implements PreDestroyOnce {
+    private final PreDestroyOnce delegate;
+
+    @Inject
+    JavaxOptionalConsumer(Optional<PreDestroyOnce> provider) {
+      this.delegate = provider.orElse(null);
+    }
+
+    public boolean isClosed() {
+      return delegate.isClosed();
+    }
+
+    public int getCount() {
+      return delegate.getCount();
+    }
   }
 
   @Singleton
   private static class PreDestroyOnceImpl implements PreDestroyOnce {
 
     private boolean closed = false;
+    private int counter = 0;
+
+    @Override
+    public boolean isClosed() {
+      return closed;
+    }
+
+    @PreDestroy
+    public void close() {
+      ++counter;
+      closed = true;
+    }
+
+    @Override
+    public int getCount() {
```

Comment:
For improved readability please call this `getPreDestroyCallCount()`

Reply:
done